src/json-rpc-socket.ts

```diff
@@ -89,33 +89,61 @@ export class JsonRpcSocket {
    * Sends a JSON-RPC request through the socket and keeps a listener open to read associated responses as they arrive.
    * Returns a close method to clean up the listener.
    */
-  subscribe(request: JsonRpcRequest, listener: (response: JsonRpcResponse) => void): { close: () => void } {
-    request.id ??= uuidv4();
+  async subscribe(request: JsonRpcRequest, listener: (response: JsonRpcResponse) => void): Promise<{
+    response: JsonRpcResponse;
+    close?: () => Promise<void>;
+   }> {
+
+    if (!request.method.startsWith('rpc.subscribe.')) {
+      throw new Error('subscribe rpc requests must include the `rpc.subscribe` prefix');
+    }
+
+    if (!request.subscribe) {
+      throw new Error('subscribe rpc requests must include subscribe options');
+    }
+
+    const subscriptionId = request.subscribe.id;
     const messageHandler = (event: { data: any }):void => {
       const jsonRpcResponse = JSON.parse(event.data.toString()) as JsonRpcResponse;
-      if (jsonRpcResponse.id === request.id) {
-        // if the incoming response id matches the request id, trigger the listener
-        return listener(jsonRpcResponse);
+      if (jsonRpcResponse.id === subscriptionId) {
+        if (jsonRpcResponse.error !== undefined) {
+          // remove the event listener upon receipt of a JSON RPC Error.
+          this.socket.removeEventListener('message', messageHandler);
+          this.closeSubscription(subscriptionId);
+        }
+        listener(jsonRpcResponse);
       }
     };
-
-    // subscribe to the listener before sending the request
     this.socket.addEventListener('message', messageHandler);
-    this.send(request);
+
+    const response = await this.request(request);
+    if (response.error) {
+      this.socket.removeEventListener('message', messageHandler);
+      return { response }
+    }
+
+    // clean up listener and create a `rpc.subscribe.close` message to use when closing this JSON RPC subscription
+    const close = async (): Promise<void> => {
+      this.socket.removeEventListener('message', messageHandler);
+      await this.closeSubscription(subscriptionId);
+    }
 
     return {
-      close: ():void => {
-        // removes the listener for this particular request
-        this.socket.removeEventListener('message', messageHandler);
-      }
-    };
+      response,
+      close
+    }
+  }
+
+  private closeSubscription(id: JsonRpcId): Promise<JsonRpcResponse> {
+    const requestId = uuidv4();
+    const request = createJsonRpcSubscribeRequest(requestId, 'rpc.subscribe.close', {}, id);
```

Comment:
Consider renaming to `createJsonRpcSubscriptionRequest`

Reply:
Done